Servlet

aleksandrzajbel25182 opened the pull request.

aleksandrzajbel25182 committed:
Change docker-compose

aleksandrzajbel25182 committed:
Add dependencies and Datasource

aleksandrzajbel25182 committed:
Сreate models

aleksandrzajbel25182 committed:
Сreate a СRUD interface

aleksandrzajbel25182 committed:
Delete HelloServlet.java

aleksandrzajbel25182 committed:
Creating a test servlet

aleksandrzajbel25182 committed:
Refactor pom

aleksandrzajbel25182 committed:
Refactor model Currencies

aleksandrzajbel25182 committed:
Create Repository

aleksandrzajbel25182 committed:
Create DatabaseConfig

aleksandrzajbel25182 committed:
Refactor name

aleksandrzajbel25182 committed:
Create sql scripts database

aleksandrzajbel25182 committed:
Delete class

aleksandrzajbel25182 committed:
Add Lisner and HikarCp

Add .xml file config

aleksandrzajbel25182 committed:
Comment out the logic

aleksandrzajbel25182 committed:
Implement a connection pool

aleksandrzajbel25182 committed:
Logging dependence

aleksandrzajbel25182 committed:
Refactoring

aleksandrzajbel25182 committed:
Do a search by currency id

Implement a currency search by id in the repository class.
Create a new servlet

aleksandrzajbel25182 committed:
Pretty Printing the Formatted JSON

aleksandrzajbel25182 committed:
Implement a servlet ExchangeRate

Implement the Exchange Rate repository and refine the model

aleksandrzajbel25182 committed:
add the ExchangeRateRepository to the listener

aleksandrzajbel25182 committed:
Rename ExchangeRatesRepository

aleksandrzajbel25182 committed:
CurrenciesRepository GetFindByCode

aleksandrzajbel25182 committed:
Implement a specific exchange rate servlet

aleksandrzajbel25182 committed:
Refactoring

aleksandrzajbel25182 committed:
Refactoring

aleksandrzajbel25182 committed:
Add a method to create a generic model method

The createEntity method that creates an instance of the model class

aleksandrzajbel25182 committed:
Fix the typo in the variable names

aleksandrzajbel25182 committed:
Test Parsing

aleksandrzajbel25182 committed:
Gson Central Bank Parser

Make a Central Bank Currency model and create a custom deserialization

aleksandrzajbel25182 committed:
BugFix

aleksandrzajbel25182 committed:
Create Currencies

aleksandrzajbel25182 committed:
working with the parser

aleksandrzajbel25182 committed:
working with the parser 2.0

aleksandrzajbel25182 committed:
Updating the exchange rate

Updates are performed by parsing the quotes of the central bank

aleksandrzajbel25182 committed:
Connect the flyway plugin

Create a version 1 migration file

aleksandrzajbel25182 committed:
Remove unnecessary classes

aleksandrzajbel25182 committed:
Remove unnecessary classes

Create a web-app module and transfer everything related to the original
project.This is done to generate a new data update module in the
background.

aleksandrzajbel25182 committed:
Сhange the path of the .war file

After creating the web-app module, change the folder path.target in docker-compose

aleksandrzajbel25182 committed:
Create a new Update Source module

Create an abstract data source and a class to implement it.
Сreate a DTO template object
Create a class to update the database.

aleksandrzajbel25182 committed:
upload changes after the update

aleksandrzajbel25182 committed:
Add comments

aleksandrzajbel25182 committed:
Hide the secrets

Refactoring pom.xml

aleksandrzajbel25182 committed:
Set it up .gitignore

aleksandrzajbel25182 committed:
Set it up .gitignore

aleksandrzajbel25182 committed:
Delete .env

aleksandrzajbel25182 committed:
Delete .idea repository

aleksandrzajbel25182 committed:
Store date properties with currency

aleksandrzajbel25182 committed:
Clean the Main class

aleksandrzajbel25182 committed:
Refactoring

aleksandrzajbel25182 committed:
Refactoring and add new currencies with a list

aleksandrzajbel25182 committed:
Incomplete update algorithm

aleksandrzajbel25182 committed:
Migrate DB

aleksandrzajbel25182 committed:
Migrate DB

aleksandrzajbel25182 committed:
Fix dependency

aleksandrzajbel25182 committed:
Return an array of currency pairs in the source

One class is enough for a pair, the ruble acts as the currency in the CB RF

aleksandrzajbel25182 committed:
Make the Collection interface support

aleksandrzajbel25182 committed:
Not a perfect update code

aleksandrzajbel25182 committed:
Create a dto transformation in an entity

aleksandrzajbel25182 committed:
Fix sql queries

aleksandrzajbel25182 committed:
Delete the list package

aleksandrzajbel25182 committed:
Updating the exchange rate

Add an update to the currency table and correct the update algorithm

aleksandrzajbel25182 committed:
Refactoring

aleksandrzajbel25182 committed:
Add the main currency RUB

aleksandrzajbel25182 committed:
Fix message output

aleksandrzajbel25182 committed:
Fix the update algorithm

Additionally, make a request to the DB to get the currency exchange id
The request must also be present in the repository

aleksandrzajbel25182 committed:
Serialize LocalDate for json

aleksandrzajbel25182 committed:
Remove the delete and save commands from CRUD

aleksandrzajbel25182 committed:
Create a currency transfer servlet

Create a model for storing funds from one currency to another.

aleksandrzajbel25182 committed:
Fix an algorithmic error during translation

aleksandrzajbel25182 committed:
Fix the value of the exchange rate

The incorrect ratio of exchange rates is stored in the database
Make the face value of the currency a unit

aleksandrzajbel25182 committed:
Correct the error in the exchange rate calculation

shining-mind commented on UpdateSource/src/main/java/update/Main.java:
```diff
@@ -0,0 +1,46 @@
+    DataSource dataSource = null;
+
+    // User name
+    String name = "postgres";
```
А как же .dotenv?

shining-mind commented on UpdateSource/src/main/java/update/Source/CBRFSource.java:
```diff
@@ -0,0 +1,108 @@
+  }
+
+
+  private ExchageRateDto createValute(Node node, LocalDate date) {
```
```suggestion
  private ExchageRateDto createExchangeRateDto(Node node, LocalDate date) {
```

shining-mind commented on UpdateSource/src/main/java/update/Source/CBRFSource.java:
```diff
@@ -0,0 +1,108 @@
+  private ExchageRateDto createValute(Node node, LocalDate date) {
+    ExchageRateDto exchageRate = new ExchageRateDto();
+    exchageRate.setDate(date);
+    if (node.getNodeType() == Node.ELEMENT_NODE) {
```
По идее надо вообще метод не вызывать если Node не ELEMENT_NODE, иначе получаем на выходе пустой ExchangeRateDto

shining-mind commented on UpdateSource/src/main/java/update/Updater.java:
```diff
@@ -0,0 +1,120 @@
+    var exchangeRatesToInsert = new ArrayList<ExchageRateDto>();
+
+    for (Map.Entry<String, Integer> entry : idByCode.entrySet()) {
+      var exchangeRateId = exchangeRateRepository.getByIdExchangeRate((entry.getValue()));
```
1) Делать в цикле много запросов в БД - это плохо, лучше сделать один запрос используя оператор `in()`
2) Непонятно почему пару мы получаем по ID валюты, ведь idByCode - это маппинг кода валюты на её ID, а чтобы получить пару нужен как минимум ID второй валюты

shining-mind commented on UpdateSource/src/main/java/update/Updater.java:
```diff
@@ -0,0 +1,120 @@
+    if (exchangeRatesToInsert != null) {
+      List<ExchangeRate> exchangeInsert = exchangeRateController.toEntity(exchangeRatesToInsert);
+      for (ExchangeRate entry : exchangeInsert) {
+        exchangeRateRepository.create(entry);
```
Лучше делать bulk insert, когда одним запросом вставляем сразу все значения

shining-mind commented on UpdateSource/src/main/java/update/Updater.java:
```diff
@@ -0,0 +1,120 @@
+
+  private HashMap<String, ExchageRateDto> getCharCodeMap(List<ExchageRateDto> exchageRates) {
+    HashMap<String, ExchageRateDto> arrayCodeMap = new HashMap<>();
+    for (ExchageRateDto currency : exchageRates) {
```
Так у нас тут все таки пары или валюты?

shining-mind commented on UpdateSource/src/main/java/update/Updater.java:
```diff
@@ -0,0 +1,120 @@
+  private HashMap<String, ExchageRateDto> getCharCodeMap(List<ExchageRateDto> exchageRates) {
+    HashMap<String, ExchageRateDto> arrayCodeMap = new HashMap<>();
+    for (ExchageRateDto currency : exchageRates) {
+      arrayCodeMap.put(currency.getCharCode(), currency);
```
Если у нас пара, то почему у нее только один код валюты. По идее должно быть два: для целевой валюты, и базовой.

shining-mind commented on UpdateSource/src/main/java/update/dto/ExchageRateDto.java:
```diff
@@ -0,0 +1,26 @@
+@Setter
+public class ExchageRateDto {
+
+  private String baseCurrency;
```
это код базовой валюты? тогда так и стоит назвать `baseCurrencyCode`

shining-mind commented on UpdateSource/src/main/java/update/dto/ExchageRateDto.java:
```diff
@@ -0,0 +1,26 @@
+
+  private String baseCurrency;
+
+  private String charCode;
```
а тут `targetCurrencyCode`

shining-mind commented on UpdateSource/src/main/java/update/dto/ExchageRateDto.java:
```diff
@@ -0,0 +1,26 @@
+
+  private double value;
+
+  private double vunitRate;
```
почему выбрал такое название?

aleksandrzajbel25182 replied:
Взял из ЦБ РФ. Понял, что плохое было решение. так как данный класс абстракция пары валют и поля должны соответствующее иметь имена.

shining-mind commented on UpdateSource/src/main/java/update/util/CurrencyController.java:
```diff
@@ -0,0 +1,23 @@
+import java.util.Map;
+import update.dto.ExchageRateDto;
+
+public class CurrencyController implements CurrencyMapToEntity<Currency> {
```
Немного непонятно почему это называется контроллером, ведь он просто преобразовывает данные.

Контроллер в шаблоне MVC (Model-View-Controller) — это компонент, который обеспечивает связь между моделью и представлением.

Он **обрабатывает действия пользователя**, проверяет полученную информацию и передаёт её модели. Контроллер определяет, как приложение будет реагировать на действия пользователя. Также контроллер может отвечать за фильтрацию данных и авторизацию.

shining-mind commented on UpdateSource/src/main/java/update/util/ExchangeRateController.java:
```diff
@@ -0,0 +1,53 @@
+import javax.sql.DataSource;
+import update.dto.ExchageRateDto;
+
+public class ExchangeRateController implements ExchangeRateDtoToEntity<ExchangeRate> {
```
Аналогичное замечание

shining-mind commented on web-app/src/main/java/exchangerate/model/Exchange.java:
```diff
@@ -0,0 +1,24 @@
+@Setter
+@NoArgsConstructor
+@AllArgsConstructor
+public class Exchange {
```
Немного не пойму зачем две модели Exchange и ExchangeRate - поясни плиз

aleksandrzajbel25182 replied:
Для расчёт перевода определённого количества средств из одной валюты в другую 
Что бы получать на выходе такое сообщение: 
{
    "baseCurrency": {
        "id": 0,
        "name": "United States dollar",
        "code": "USD",
        "sign": "$"
    },
    "targetCurrency": {
        "id": 1,
        "name": "Australian dollar",
        "code": "AUD",
        "sign": "A€"
    },
    "rate": 1.45,
    "amount": 10.00,
    "convertedAmount": 14.50
}
Или такое сообщение надо было собирать в самом servlet?

shining-mind commented on web-app/src/main/java/exchangerate/repository/CurrenciesRepository.java:
```diff
@@ -0,0 +1,166 @@
+    List<Currency> currencies = new ArrayList<>();
+
+    try (Connection connection = dataSource.getConnection();
+        PreparedStatement statement = connection.prepareStatement(GET_ALL_CURRENCIES)) {
```
Ну,  кстати тут можно сразу запрос выполнять без prepared statement, так как мы в него не подставляем данные

shining-mind commented on web-app/src/main/java/exchangerate/repository/CurrenciesRepository.java:
```diff
@@ -0,0 +1,166 @@
+    } catch (SQLException e) {
+      throw new RuntimeException(e);
+    }
+    return currency;
```
Как потребитель определить, что валюта не найдена, если всегда возвращается объект?

shining-mind commented on web-app/src/main/java/exchangerate/repository/CurrenciesRepository.java:
```diff
@@ -0,0 +1,166 @@
+    StringBuilder sql = new StringBuilder(
+        "SELECT id, charcode FROM currencies WHERE charcode IN (");
+
+    for (int i = 0; i < collection.size(); i++) {
```
вместо цикла можно сделать map в массив строк из `?` и потом `join`:

```java
sql.append(String.join(", ", arrayOfStrings));
```

Но цикл тоже окей, просто чуть более шумно

shining-mind commented on web-app/src/main/java/exchangerate/servlets/ExchangeRateServlet.java:
```diff
@@ -0,0 +1,37 @@
+    var baseCurrency = currenciesCodes.substring(0, 3);
+    var targetCurrency = currenciesCodes.substring(3, 6);
+    PrintWriter writer = resp.getWriter();
+    var message = JsonConvert.jsonConvert(exchangeRateRepository.finByCode(baseCurrency, targetCurrency));
```
```suggestion
    var message = JsonConvert.jsonConvert(exchangeRateRepository.findByCode(baseCurrency, targetCurrency));
```

shining-mind commented on web-app/src/main/java/exchangerate/servlets/ExchangeServlet.java:
```diff
@@ -0,0 +1,93 @@
+
+  }
+
+  private BigDecimal getRate(String from, String to) {
```
Вот это надо перенести в ExchangeService, логика по расчёт суммы обмена - не должна зависеть от контекста HTTP

Таким образом этот сервис можно будет использовать в любом контексте: email рассылка, отправка уведомления, HTTP, CLI и т.п.

shining-mind commented on web-app/src/main/resources/db/migration/V2__Insert.sql:
```diff
@@ -0,0 +1,23 @@
+    (9, 'JPY', 'Yen'),
+    (10, 'CNY', 'Yuan Renminbi');
+
+INSERT INTO exchangerates (id, basecurrencyid, targetcurrencyid, rate, date)
```
Ну, вот пары обмена наверное не стоит в миграцию добавлять, так как их всегда можно получить из ЦБ РФ.

aleksandrzajbel25182 committed:
Hide the secrets

Get data from env by duplicating it into a module

aleksandrzajbel25182 committed:
Refactoring

aleksandrzajbel25182 committed:
Fix the typo of the name

Add minor documentation

aleksandrzajbel25182 committed:
Correct the name of the classes to the appropriate

aleksandrzajbel25182 committed:
Refactoring

aleksandrzajbel25182 committed:
Extracted exchange some logic into separate class

aleksandrzajbel25182 committed:
Remove currency pairs from migration

aleksandrzajbel25182 committed:
Refactoring

aleksandrzajbel25182 committed:
Send data to the database in batches

Batch processing for currency pairs

aleksandrzajbel25182 committed:
Make a batch transfer to the database

aleksandrzajbel25182 committed:
Refactoring

aleksandrzajbel25182 committed:
Fix the update algorithm

aleksandrzajbel25182 committed:
Refactoring

aleksandrzajbel25182 committed:
Fix update

aleksandrzajbel25182 committed:
Refactoring

shining-mind commented on UpdateSource/src/main/java/update/Main.java:
```diff
@@ -0,0 +1,42 @@
+    DataSource dataSource = null;
+    PGSimpleDataSource ds = new PGSimpleDataSource();
+    Dotenv dotenv = Dotenv.configure()
+        .directory("UpdateSource/")
```
А у тебя .dotenv не в корне проекта?

shining-mind commented on UpdateSource/src/main/java/update/Updater.java:
```diff
@@ -0,0 +1,118 @@
+    var exchangeRatesToUpdate = new ArrayList<Pair<Integer, ExchangeRateDto>>();
+    var exchangeRatesToInsert = new ArrayList<ExchangeRateDto>();
+    // 3.1 Creating a Map with currency pairs and ExchangeRateDto
+    Map<Pair<Integer, Integer>, ExchangeRateDto> pairsCharCodesMap = new HashMap<>();
```
Вот тут важный момент, почему пара может быть ключом в HashMap. Чтобы значение использовать в качестве ключа у него должны быть реализованы два метода: equals и hashCode, таким образом HashMap сможет сравнивать и проверять вхождение.

В доке видно, что эти методы у пары есть https://docs.oracle.com/javase/9/docs/api/javafx/util/Pair.html

Что из этого следует? При необходимости ты сможешь сам реализовать структуру, которую можно будет использовать в качестве ключа в HashMap

shining-mind commented on UpdateSource/src/main/java/update/dto/ExchangeRateDto.java:
```diff
@@ -0,0 +1,52 @@
+  /**
+   * @value vunitRate - The rate for 1 unit of currency
+   */
+  private double vunitRate;
```
По описанию не могу понять чем отличается это поле от `rate`. Точно не забыл удалить?

aleksandrzajbel25182 replied:
Цб рф же дает курс на номинал к примеру у турецкой лиры номинал: 10 TRY rate = 26,2382 
vunitRate же на единицу показывает курс 2,6238

shining-mind replied:
Тогда надо описание обновить, потому что оно сейчас одинаковое и как-будто оба поля показывают курс к 1 единице валюты

shining-mind commented on web-app/src/main/java/exchangerate/repository/CurrenciesRepository.java:
```diff
@@ -0,0 +1,178 @@
+        preparedStatement.setString(2, entity.getFullName());
+        preparedStatement.addBatch();
+      }
+      preparedStatement.executeBatch();
```
В целом - это уже лучше, потому что будет отправлен один запрос, но в нем будет много команд.

А можно сделать ещё круче:

`INSERT INTO table VALUES (1, 2), (3, 4)`

https://stackoverflow.com/questions/20815028/how-do-i-insert-multiple-values-into-a-postgres-table-at-once

Думаю это должно работать чуть быстрее, так как insert один, но много значений, и базе данных не надо парсить каждый запрос отдельно. Можешь даже для интереса написать скрипт и попробовать вставить 1000 значений через executeBatch и одни запросом с много Values.

shining-mind commented on web-app/src/main/java/exchangerate/repository/CurrenciesRepository.java:
```diff
@@ -0,0 +1,178 @@
+      preparedStatement.setString(1, entity.getCode());
+      preparedStatement.setString(2, entity.getFullName());
+      preparedStatement.setInt(3, entity.getId());
+      preparedStatement.executeUpdate();
```
Вот здесь абсолютно валидно, так как в SQL нет UPDATE для множества значений. 

Хотя этого можно добиться INSERT'ом с ON CONFLICT DO UPDATE (но это скорее хак). Такой хак меня выручал когда нужно было обновлять много значений - это работало существенно быстрее, чем много UPDATE'ов. Правда у меня драйвер не поддерживал batch запросы, возможно с ними выигрышь был бы не такой существенный.

Вот сам хак:

https://www.postgresqltutorial.com/postgresql-tutorial/postgresql-upsert/

Лучше иметь такой туз в рукаве и знать о нём. По сути это эмуляция операции UPSERT - вставить если нет или обновить если есть.

shining-mind commented on web-app/src/main/java/exchangerate/utils/ExchangeService.java:
Не уверен, что сервисы стоит в utils размещать. Сервис - это про бизнес логику приложения, их надо в свою директорию.

shining-mind commented on UpdateSource/src/main/java/update/Updater.java:
```diff
@@ -0,0 +1,118 @@
+        uniquePairs);
+
+    // 4. Divide it into lists for updating and adding
+    for (Pair<Integer, Integer> pair : uniquePairs) {
```
Попробуй эту часть алгоритма переписать на UPSERT, чтобы не надо было формировать два массива: на вставку и обновление. Про UPSERT я написал ниже в репозитории.

aleksandrzajbel25182 committed:
Add the reWriteBatchedInserts config property

aleksandrzajbel25182 committed:
Refactoring file locations

aleksandrzajbel25182 committed:
Fix the path to the .env directory file

aleksandrzajbel25182 committed:
Fix bug

aleksandrzajbel25182 committed:
Refactoring

aleksandrzajbel25182 committed:
Refactoring and documenting dto and source

aleksandrzajbel25182 committed:
Use upsert to update currency pairs

To use the conflict on a currency pair, you need to add 3 migrations
Use UNIQUE for currency pairs

aleksandrzajbel25182 committed:
Set up new migrations

aleksandrzajbel25182 committed:
Implement the upsert algorithm for currencies

shining-mind commented on UpdateSource/src/main/java/update/Updater.java:
```diff
@@ -40,8 +40,21 @@ public Updater(DataSource dataSource, CurrencyExchangeRateSource source, LocalDa
+
+    // TODO Below is an implementation of the algorithm without using UPSERT.
 
+    /* The algorithm for updating and adding without using upsert
```
Вынеси в отдельный метод. Решение о том какой алгоритм применять делай по наличию ENV переменной. Можешь задать ENV переменную `DB_SUPPORTS_UPSERT`, и если она true, то выполняй upsert, иначе старый алгоритм.

Такой подход часто используется на практике и называется фича флагом (feature flag).

shining-mind commented on web-app/src/main/java/exchangerate/repository/CurrenciesRepository.java:
```diff
@@ -150,6 +149,26 @@ public void createBatch(List<Currency> entities) {
   }
 
+  public void upsert(List<Currency> entities) {
+    StringBuilder sql = new StringBuilder(
```
Интересно он такой запрос тоже перепишет под множественную вставку, у меня есть сомнения. Надо бы проверить - думаю у драйвера должен быть дебаг режим, где видно какие запросы уходят в БД.

aleksandrzajbel25182 replied:
![image](https://github.com/user-attachments/assets/fb868ea4-03c6-4059-993b-fac0f297e68e)
Посмотрел оказывается в логах контейнера прописывается запрос. Ну я насколько понял он под множественную вставку пошел

shining-mind commented on web-app/src/main/java/exchangerate/repository/ExchangeRateRepository.java:
```diff
@@ -198,6 +194,27 @@ public void updateBatch(List<ExchangeRate> entities) {
+    StringBuilder sql = new StringBuilder(
+        "INSERT INTO exchangerates (basecurrencyid,targetcurrencyid,rate,date) "
+            + "VALUES (?,?,?,?) "
+            + "ON CONFLICT(basecurrencyid,targetcurrencyid) DO UPDATE "
```
А вот тут условие неверное - конфликтом является кортеж из 3 элементов: 1 и 2 валюта, а также дата.

Иначе ты будешь всегда перезаписывать все пары или только первую встретившуюся, тут не уверен. Также тебе нужно создать уникальный индекс по 3 столбцам (basecurrencyid, targetcurrencyid, date).

По факту это и есть первичный ключ отношения, но для эффективности мы используем суррогатный первичный ключ (инкрементирующийся id).

aleksandrzajbel25182 committed:
Fix

aleksandrzajbel25182 committed:
Feature flag

aleksandrzajbel25182 committed:
Refactoring configuration files

aleksandrzajbel25182 committed:
Add common classes to the module

aleksandrzajbel25182 committed:
Add currency pair subscriptions to the database

aleksandrzajbel25182 committed:
Add the subscription entity

aleksandrzajbel25182 committed:
Fix name

aleksandrzajbel25182 committed:
Replace the create with upsert method

aleksandrzajbel25182 committed:
Fix migration files

aleksandrzajbel25182 committed:
Add a servlet to subscriptions

aleksandrzajbel25182 committed:
Put the JsonConverter in the core module

aleksandrzajbel25182 committed:
Configure the pom

aleksandrzajbel25182 committed:
Fix import

aleksandrzajbel25182 committed:
Make a notification about subscriptions

aleksandrzajbel25182 committed:
Correct the sql-request

aleksandrzajbel25182 committed:
Refactoring

aleksandrzajbel25182 committed:
Document the entity package

aleksandrzajbel25182 committed:
Wrap it in an optional value where it can be null

aleksandrzajbel25182 committed:
Handling errors in currencies

aleksandrzajbel25182 committed:
Handle errors

aleksandrzajbel25182 committed:
Refactoring error handling

aleksandrzajbel25182 committed:
Fix bug

aleksandrzajbel25182 committed:
Validation part1

aleksandrzajbel25182 committed:
Validation data

aleksandrzajbel25182 committed:
Fix

aleksandrzajbel25182 committed:
Document the core module

aleksandrzajbel25182 committed:
Update the status after sending the notification

aleksandrzajbel25182 committed:
Add Encoding  Filter

shining-mind commented on core/src/main/java/com/entities/Exchange.java:
```diff
@@ -0,0 +1,50 @@
+  /**
+   * The result of converting the amount of the base currency based on the specified exchange rate.
+   */
+  BigDecimal convertToAmount;
```
Здесь название похоже на название метода, а не свойства. Потому что указан глагол convert, если перевести дословно convert to amount - конвертировать в сумму

shining-mind commented on core/src/main/java/com/entities/ExchangeRate.java:
```diff
@@ -0,0 +1,52 @@
+@Setter
+@AllArgsConstructor
+@NoArgsConstructor
+public class ExchangeRate {
```
Вот эта модель, очень похожа на предыдущую. На сколько понимаю предыдущая модель используется для того, чтобы сформировать ответ в API. Возможно ту модель стоит назвать ExchangeDTO

shining-mind commented on core/src/main/java/com/entities/Subscriptions.java:
```diff
@@ -0,0 +1,57 @@
+@Setter
+@AllArgsConstructor
+@NoArgsConstructor
+public class Subscriptions {
```
А почему во множественном числе. У нас же здесь одна подписка.

shining-mind commented on core/src/main/java/com/interfaces/CrudRepository.java:
```diff
@@ -0,0 +1,60 @@
+   *
+   * @param entities the list of entities to be upserted
+   */
+  void upsert(List<T> entities);
```
Я бы поддержал возможность делать upsert и для одной сущности, для этого тебе надо задать две сигнатуры метода.

shining-mind commented on core/src/main/java/com/interfaces/CrudRepository.java:
```diff
@@ -0,0 +1,60 @@
+   *
+   * @param entities the list of entities to be created
+   */
+  void createBatch(List<T> entities);
```
Можно использовать название `create` - это Параметрический полиморфизм. Вот что говорит ChatGPT на эту тему:

Да, в Java можно объявлять методы с одинаковыми именами, но с различными сигнатурами в одном и том же интерфейсе. Это концепция называется перегрузкой методов. В вашем случае, переименование метода createBatch в create, который принимает параметр List<T>, сохраняя при этом другой метод create, который принимает один экземпляр T, допускается в Java. Вот как будет выглядеть интерфейс:

```java
public interface CrudRepository<T> {
    /**
     * Создает новую сущность в базовом хранилище данных.
     *
     * @param entity сущность, которую нужно создать
     */
    void create(T entity);

    /**
     * Создает пакет сущностей в базовом хранилище данных.
     *
     * @param entities список сущностей для создания
     */
    void create(List<T> entities);
}
```
### Важные аспекты перегрузки методов

1. Ясность типов: Перегрузка методов иногда может уменьшить ясность кода, особенно если типы параметров тесно связаны (например, перегрузка метода для List<T> и для одного T может быть понятной, но перегрузка для int и Integer может быть запутанной).

2. Удобство API: Перегрузка не должна сбивать пользователей API с толку. Должно быть сразу очевидно, какой метод вызывается на основании передаваемых аргументов. В вашем случае, так как типы параметров (T против List<T>) явно различаются, это способствует интуитивному использованию.

3. Совместимость и гибкость: Перегруженные методы позволяют различные способы использования одной и той же функциональности (в вашем случае создание одной или нескольких сущностей), что может сделать API более гибким и легким в использовании без необходимости запоминания многих названий методов.

### Является ли это хорошей практикой?

Перегрузка в данном случае может рассматриваться как хорошая практика, если она соответствует следующим условиям:

- Явная ясность: Методы выполняют по сути одну и ту же операцию (создание сущностей), где единственное различие - это количество. Это хорошо соответствует принципу интерфейса, определяющего согласованный набор операций.

- Согласованность в реализации: Важно обеспечить, чтобы все реализации этого интерфейса соответствовали семантике обоих перегруженных методов корректно, не внося непредвиденных побочных эффектов или последствий для производительности.

- Документация: Предоставьте четкую документацию (как вы начали делать) для каждой вариации метода, чтобы пользователи вашего интерфейса имели четкое понимание, когда использовать каждый метод.

Однако, если поведение методов значительно различается помимо обработки одиночных или множественных элементов (например, если один из методов начинает включать разные характеристики транзакций или обработки ошибок), возможно, будет лучше сохранить для них разные названия, чтобы отразить эти различия.

В заключение, переименование createBatch в create в вашем интерфейсе для целей перегрузки полностью приемлемо в данном контексте, при условии, что реализация и документация четко поддерживают этот выбор дизайна.

shining-mind commented on core/src/main/java/com/interfaces/NotificationSender.java:
```diff
@@ -0,0 +1,27 @@
+   * notification. During implementation, it is necessary to ensure that the notification is
+   * successfully sent to the intended recipient(s).</p>
+   */
+  void send();
```
Не, уверен что void это хорошее решение. Здесь стоило бы использовать [Future](https://docs.oracle.com/en/java/javase/17/docs/api/java.base/java/util/concurrent/Future.html)

В некоторых случаях можно пренебрегать результатом отправки, но обычно мы хотим предоставить возможность потребителю нашего API обработать результат (чтобы в случае ошибки попробовать ещё раз).

shining-mind commented on core/src/main/java/com/repository/CurrenciesRepository.java:
```diff
@@ -41,77 +56,89 @@ public class CurrenciesRepository implements CrudRepository<Currency> {
   @Override
-  public List<Currency> get() {
-    List<Currency> currencies = new ArrayList<>();
+  public void create(Currency entity) {
```
Вообще, было бы неплохо, чтобы метод возвращал созданную модель БД, ну либо патчил объект, который мы ему передаем на вход (добавлял ID)

shining-mind replied:
https://stackoverflow.com/questions/1915166/how-to-get-the-insert-id-in-jdbc

shining-mind commented on core/src/main/java/com/repository/CurrenciesRepository.java:
```diff
@@ -41,77 +56,89 @@ public class CurrenciesRepository implements CrudRepository<Currency> {
   @Override
-  public Currency findById(int id) {
-    Currency currency = new Currency();
+  public void createBatch(List<Currency> entities) {
```
Тут аналогичное предложение

shining-mind commented on core/src/main/java/com/repository/CurrenciesRepository.java:
```diff
@@ -169,21 +196,75 @@ public void upsert(List<Currency> entities) {
+   * `Optional` if not found
+   * @throws RuntimeException if an SQL exception occurs during the database query
+   */
+  public Optional<Currency> findByCode(String hasCode) {
```
А почему в findById не сделал optional? Там же тоже возможна ситуация, когда по ID не получится найти

shining-mind commented on core/src/main/java/com/repository/ExchangeRateRepository.java:
```diff
@@ -159,25 +211,50 @@ public void createBatch(List<ExchangeRate> entities) {
+   * `Optional` if not found
+   * @throws RuntimeException if an SQL exception occurs during the database query
+   */
+  public Optional<ExchangeRate> finByCode(String baseCurrency, String targetCurrency) {
```
```suggestion
  public Optional<ExchangeRate> findByCode(String baseCurrency, String targetCurrency) {
```

shining-mind commented on core/src/main/java/com/services/Notification.java:
```diff
@@ -0,0 +1,53 @@
+  /**
+   * The base currency ID for the exchange rate.
+   */
+  private Currency baseCurrencyId;
```
Может просто в notification добавить модель ExchangeRate, чтобы не дублировать свойства? Композиция объектов это удобно

shining-mind commented on core/src/main/java/com/services/NotificationSenderService.java:
```diff
@@ -0,0 +1,110 @@
+ * @author Александр Зайбель
+ * @version 1.0
+ */
+public class NotificationSenderService implements NotificationSender {
```
Не совсем корректно делать сервис, который расширяет интерфейс для отправки. Сервис он сам по себе, но он как раз должен использовать некий транспорт для отправки, который реализует интерфейс NotificationSender. То есть для отправки тебе надо сделать отдельный класс

shining-mind commented on core/src/main/java/com/services/NotificationSenderService.java:
```diff
@@ -0,0 +1,110 @@
+
+        OutputStream outputStream = connection.getOutputStream();
+
+        outputStream.write(notification.getValue().getBytes());
```
Я бы выполнял запрос асинхронно, чтобы не блокировать поток.

shining-mind commented on core/src/main/java/com/services/NotificationSenderService.java:
```diff
@@ -0,0 +1,110 @@
+   */
+  @Override
+  public void send() {
+    List<Subscriptions> subscriptions = subscriptionsRepository.getByStatus("not sent");
```
Надо добавить критическую секцию. Самый просто способ: это создать траназакцию, извлечь все записи с запросом SELECT FOR UPDATE, в конце работы закомиттить транзакцию.

Также я рекомендовал бы за один проход обрабатывать не более 10-20 записей.

shining-mind commented on updateSource-app/src/main/java/update/Updater.java:
```diff
@@ -51,12 +58,11 @@ public void updateExchageRate() {
-    // TODO Below is an implementation of the algorithm without using UPSERT.
-
-    /* The algorithm for updating and adding without using upsert
+  private void baseUpdate(List<ExchangeRateDto> exchangeRates) {
```
Здесь стоит добавить описание методу, что означает baseUpdate

shining-mind commented on web-app/src/main/java/exchangerate/service/ExchangeService.java:
```diff
@@ -16,26 +18,26 @@ public ExchangeService(ExchangeRateRepository exchangeRateRepository) {
-    ExchangeRate exchangeRate = exchangeRateRepository.finByCode(from, to);
-    if (exchangeRate != null) {
-      return exchangeRate.getRate();
+    Optional<ExchangeRate> exchangeRate = exchangeRateRepository.finByCode(from, to);
```
```suggestion
    Optional<ExchangeRate> exchangeRate = exchangeRateRepository.findByCode(from, to);
```

aleksandrzajbel25182 committed:
Refactor of the Exchange and Subscriptions classes

aleksandrzajbel25182 committed:
Refactoring CRUD operation

aleksandrzajbel25182 committed:
Update the findById method interface to optional

aleksandrzajbel25182 committed:
Refactoring Notification

Get rid of code duplication by using object composition

aleksandrzajbel25182 committed:
Refactoring sending notifications

aleksandrzajbel25182 committed:
Add the Updater documentation

aleksandrzajbel25182 committed:
Add the Updater documentation

aleksandrzajbel25182 committed:
Extend the functionality of the creation method

Enhanced create method: return generated keys or correct input objects
